omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java

```diff
@@ -37,71 +37,76 @@
+
+    /**
+     * @see javax.servlet.Filter#init(javax.servlet.FilterConfig)
+     */
+    @Override
+    public void init(FilterConfig arg0) throws ServletException {
+        log.debug("Initializing REST WS Authorization filter");
+    }
+
+    /**
+     * @see javax.servlet.Filter#destroy()
+     */
+    @Override
+    public void destroy() {
+        log.debug("Destroying REST WS Authorization filter");
+    }
+
+    /**
+     * @see javax.servlet.Filter#doFilter(javax.servlet.ServletRequest,
+     *      javax.servlet.ServletResponse, javax.servlet.FilterChain)
+     */
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException,
+            ServletException {
+
+        // check the IP address first.  If its not valid, return a 403
+        if (!RestUtil.isIpAllowed(request.getRemoteAddr())) {
+            // the ip address is not valid, set a 403 http error code
+            HttpServletResponse httpresponse = (HttpServletResponse) response;
+            httpresponse.sendError(HttpServletResponse.SC_FORBIDDEN, "IP address '" + request.getRemoteAddr()
+                    + "' is not authorized");
+            return;
+        }
+
+        // skip if the session has timed out, we're already authenticated, or it's not an HTTP request
+        if (request instanceof HttpServletRequest) {
+            HttpServletRequest httpRequest = (HttpServletRequest) request;
+            if (httpRequest.getRequestedSessionId() != null && !httpRequest.isRequestedSessionIdValid()) {
+                HttpServletResponse httpResponse = (HttpServletResponse) response;
+                httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Session timed out");
+            }
+
+            if (!Context.isAuthenticated()) {
+                String basicAuth = httpRequest.getHeader("Authorization");
+                if (basicAuth != null) {
+                    // this is "Basic ${base64encode(username + ":" + password)}"
+                    try {
+                        basicAuth = basicAuth.substring(6); // remove the leading "Basic "
+                        String decoded = new String(Base64.decodeBase64(basicAuth), Charset.forName("UTF-8"));
+                        String[] userAndPass = decoded.split(":");
+                        Context.authenticate(userAndPass[0], userAndPass[1]);
+                        if (log.isDebugEnabled())
+                            log.debug("authenticated " + userAndPass[0]);
+                    } catch (Exception ex) {
+                        // This filter never stops execution. If the user failed to
+                        // authenticate, that will be caught later.
+                    }
+                } else {
+                    HttpServletResponse httpResponse = (HttpServletResponse) response;
+                    httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Basic header does not have a provided credential");
```

Comment:
In this case, `basicAuth` is `null`, which means that the `Authorization` header is unset. This should return the 401 error indicated in step 2.1.

```suggestion
                    HttpServletResponse httpResponse = (HttpServletResponse) response;
                    httpResonse.setHeader("WWW-Authenticate", "Basic, OpenMRS-Cookie")
                    httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid credentials provided");
```